File: backend/app/models/user_model.py
from enum import Enum
from datetime import datetime
from sqlmodel import SQLModel, Field
from typing import Optional


class UserRole(str, Enum):
    CUSTOMER = "CUSTOMER"
    DEALER = "DEALER"
    USER = "USER"
    ADMIN = "ADMIN"
    ROOT_ADMIN = "ROOT_ADMIN"


class User(SQLModel, table=True):
    id: Optional[int] = Field(
        default=None,
        primary_key=True
    )

    name: str
    email: str = Field(unique=True, index=True)
    password: str
    phone: Optional[str] = None
    role: UserRole = Field(default=UserRole.CUSTOMER)
    city: Optional[str] = None
    avatar_url: Optional[str] = None
    email_verified: bool = Field(default=False)
    email_verified_at: Optional[datetime] = None
    created_at: datetime = Field(default_factory=datetime.utcnow)

File: frontend/lib/modules/auth/domain/models/user_model.dart
enum UserRole {
  customer,
  dealer,
  admin;

  String get displayName {
    switch (this) {
      case UserRole.customer:
        return 'Customer';
      case UserRole.dealer:
        return 'Dealer';
      case UserRole.admin:
        return 'Admin';
    }
  }

  static UserRole fromString(String? role) {
    switch (role?.toLowerCase()) {
      case 'dealer':
        return UserRole.dealer;
      case 'admin':
        return UserRole.admin;
      case 'customer':
      default:
        return UserRole.customer;
    }
  }
}

class UserModel {
  final String id;
  final String email;
  final String fullName;
  final UserRole role;
  final String? phone;
  final String? avatarUrl;

  const UserModel({
    required this.id,
    required this.email,
    required this.fullName,
    required this.role,
    this.phone,
    this.avatarUrl,
  });

  factory UserModel.fromJson(Map<String, dynamic> json) {
    return UserModel(
      id: json['id'] as String? ?? '',
      email: json['email'] as String? ?? '',
      fullName: json['full_name'] as String? ?? json['name'] as String? ?? 'User',
      role: UserRole.fromString(json['role'] as String?),
      phone: json['phone'] as String?,
      avatarUrl: json['avatar_url'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'email': email,
      'full_name': fullName,
      'role': role.name,
      'phone': phone,
      'avatar_url': avatarUrl,
    };
  }

  UserModel copyWith({
    String? id,
    String? email,
    String? fullName,
    UserRole? role,
    String? phone,
    String? avatarUrl,
  }) {
    return UserModel(
      id: id ?? this.id,
      email: email ?? this.email,
      fullName: fullName ?? this.fullName,
      role: role ?? this.role,
      phone: phone ?? this.phone,
      avatarUrl: avatarUrl ?? this.avatarUrl,
    );
  }
}
